scripts/initialize-build-host.sh

```diff
@@ -153,11 +153,46 @@ then
     fi
 
     # --------------------------------------------------------------------------
-    # Populate build host.
+    # Check nested VM health and reset it, if needed.
     # --------------------------------------------------------------------------
 
     login="$(cat $HOME/proxy-target.txt)"
 
+    if [ -z "$login" ]
+    then
+        echo "Nested VM not started, aborting the build"
+        exit 1
+    fi
+
+    if $RSH $login true
+    then
+	:
+    else
+        echo "Failed to SSH to nested VM, probably it's hanging, resetting it"
+        VM_id="$(sudo virsh list | cut -d' ' -f 2 | sed 's/[^0-9]//g;/^$/d')"
+        sudo virsh reset $VM_id
```

Comment:
I'm not sure this will work correctly. IIRC the `proxy-target.txt` mechanism is shared with the exotic platforms, and they are not running in a VM. `proxy-target.txt` itself is not a guarantee that we are dealing with a nested VM.

Why is this needed in the build script, and not in the launch script for the VM? Once you've verified it's up, it won't go down, will it? Or is this because it goes down mid-build?

Reply:
Yes. In the https://ci.cfengine.com/job/testing-enterprise-master/configure at the top of "Build" script, there is this command:
{code}
test -f $WORKSPACE/mender-qa/scripts/initialize-build-host.sh && \
    .   $WORKSPACE/mender-qa/scripts/initialize-build-host.sh
{code}
Dunno who put it, but yes this script runs on each and every build.

And for example here: https://ci.cfengine.com/job/testing-enterprise-master/343/label=PACKAGES_x86_64_linux_debian_4/console
we have a typical case of failure due to nested VM hang:
{code}
01:16:29 + cat /home/jenkins/proxy-target.txt
01:16:29 + login=jenkins@192.168.122.210
01:16:29 + rsync --delete -czrlpt -T /tmp -e ssh -o BatchMode=yes /tmp/hudson2811054020807905892.sh jenkins@192.168.122.210:commands-from-proxy.sh
01:16:32 ssh: connect to host 192.168.122.210 port 22: No route to host
01:16:32 rsync: connection unexpectedly closed (0 bytes received so far) [sender]
01:16:32 rsync error: error in rsync protocol data stream (code 12) at io.c(226) [sender=3.1.1]
01:16:32 Build step 'Execute shell' marked build as failure
{code}

At this point we can reset the nested VM, and after that it will be happily working.

When we're talking about proxying to exotic platforms, I believe if any of them is hang, our build process is doomed anyway. Probably we could check that we're dealing with nested VM, and not with proxy-to-exotic, before failing, to print a nicer error message.

Reply:
Yes, maybe that would be better, verify that this truly is a nested VM. And if possible it would be nice if this code could be moved out of `initalize-build-host.sh` and into for example `nested-vm.sh`, or some other file, and `initalize-build-host.sh` just calls that one. The reason is that this file is used on *all* build hosts, including Mender, so it would be nice if this very nested-VM-specific thing was stowed away somewhere else.

> ```
> test -f $WORKSPACE/mender-qa/scripts/initialize-build-host.sh && \
>    .   $WORKSPACE/mender-qa/scripts/initialize-build-host.sh
> ```

Yes, this is run on all jobs to make sure to transport the commands to the proxy target and run them there instead, allowing the build recipe to remain identical for all hosts.

Reply:
Right. Then, maybe `nested-vm.sh` can create `nested-vm-check.sh` script with above code, and testing-enterprise-master Jenkins job can check for this file and execute it, if it's there?

Reply:
Moving it into another file is complicated, since I'm not quite sure in which dir it is. As a comment below in this file says, provisioned hosts will have "mender-qa" repo in HOME, while permanent hosts will have it in the WORKSPACE instead.

Reply:
Right, and because the file is sourced, it's even more difficult... Ok, we drop that, just put it in a function then, to keep it tidy.

Reply:
moved all nested VM stuff into a separate function! Ok to merge?